Change `docker volume create` to non-idempotent

yongtang opened the pull request:
**\- What I did**

This fix tries to address the issue raised in #16068 where non-idempotency is desired for `docker volume create`.

**\- How I did it**

Previously `docker volume create` will silently ignore in case a volume with the same name has already been created with the same driver (idempotency). This fix made the change so that an `already exist` error is returned together with the already created volume `v` in `Create/CreateWithRef` func. 

(So a check of return error type is needed if the caller wants idempotency instead)

**\- How to verify it**

Tests has been added to cover the chagnes.

**\- Description for the changelog**

An error will be returned for `docker volume create` in case a volume with the same name has been created by the same driver.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #16068.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

thaJeztah commented:
Oh! I just noticed that this probably does the same as https://github.com/docker/docker/pull/27346 (haven't looked at the implementation yet)

Also we should take into account that we now store the options that were used when a volume was created, so perhaps `volume create` does not have to produce an error if all options are exactly the same

allencloud commented:
Oh, @thaJeztah 
To be honest, maybe we should be more careful on what you suggest that `same name with same options`.

Here is my own thought.
Since if a user executes `docker volume create --name aaa` twice(same name same options), probably he indeed needs two different volumes.(maybe he forgot he created before or someone else did that) If dockerd make the second action succeed, afterwards the volume may be polluted. Its damage will be exposed.

Maybe needs more discussion in this part.

cpuguy83 commented on volume/store/store.go:
```diff
@@ -269,14 +273,16 @@ func (s *VolumeStore) create(name, driverName string, opts, labels map[string]st
 		}
-		return v, nil
+		// In case a volume has already been created, we will return an error (non-idempotency)
+		return v, errAlreadyCreated
```
I'm not a huge fan of this approach.

Ultimately the reason the code is as it is was to support the legacy auto-create/use for `docker run -v <name>:/foo`.
Meanwhile this pattern is racey:

```
Get(foo) || Create(foo)
```

Maybe we could have `Get` accept a callback that gets run at the end of the fn while the store is still locked.

yongtang commented:
@cpuguy83 @allencloud @thaJeztah I made some changes to the PR. Now instead of checking the error, the `CreateWithRef` will carry a arg of `quiet bool`. If `quiet == true`, then the behavior will be the same as before. If `quiet == false`, then an error `already exist` will be returned. Please take a look and let me know if this will fit the requirement.

cpuguy83 commented on volume/store/store.go:
```diff
@@ -247,15 +249,17 @@ func (s *VolumeStore) CreateWithRef(name, driverName, ref string, opts, labels m
-	return s.CreateWithRef(name, driverName, "", opts, labels)
+// In case a volume with the name has already been created, an errAlreadyCreated error will be
+// returned along with the already created volume, if `quiet` is not true
+func (s *VolumeStore) Create(name, driverName string, opts, labels map[string]string, quiet bool) (volume.Volume, error) {
```
I don't think this is a good solution.
Bool parameters are never a good solution.

yongtang replied:
Thanks @cpuguy83. No sure about the best way to handle the case. Any suggestions?

cpuguy83 replied:
@yongtang I think it would be better to have a `CreateIdempotent` than a bool arg... or a set of functional arguments (http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)

yongtang committed:
Change `docker volume create` to non-idempotent

This fix tries to address the issue raised in 16068 where
non-idempotency is desired for `docker volume create`.

Previously `docker volume create` will silently ignore in case
a volume with the same name has already been created with the
same driver (idempotency). This fix made the change so that
an `already exist` error is returned together with the already
created volume `v` in `Create/CreateWithRef` func.

The change is to add functional options to `Create` so that
caller has the option to return error if a volume has been
found.

Tests has been added to cover the chagnes.

This fix fixes 16068.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cpuguy83. The PR has been updated with functional options. Please take a look.

cpuguy83 reviewed (commented):
Thanks, this is definitely much cleaner.

Right now this is implemented more like a callback.
I wonder if it would be better to do more like a config struct:

ie, the functional argument signature would look like `func(*CreateConfig)` where in the argument you set options (like "Ref" and idempotency handling).

Then we always call `for _, arg := range args { f(cfg) }` at the beginning of `Create` and check the options struct in the function body as needed.

yongtang commented:
@cpuguy83 One issue I am encountering is that, idempotency check is done at the end of `Get` while setting Refs is done at the end of `Create`.

We may need to pass a flag to indicate if the volume is obtained from the existing one or created new, before invoking an array of option functionals.

I will spend more time thinking about if there is a better way to handle the scenario.

cpuguy83 commented:
@yongtang Maybe we can just have an `CreateIdempotent` for `docker run -v` and a normal `Create` for `docker volume create`?

This will definitely need to be versioned in the API, though... and probably go through the deprecation policy.

yongtang commented:
Thanks @cpuguy83 for the suggestion. Let me take a look and update the PR accordingly.

yongtang commented:
Thanks @cpuguy83. The PR has been updated to take API version into consideration. Also added the CreateIdempotent function. Please take a look.

cpuguy83 commented:
@yongtang Thanks! Seems like we're getting closer.
Now what if `volumes.Create` returned something like an `ErrAlreadyExists` when name and driver are the same. Then at the API we can determine how to respond to the client.

Namely I think in older API's we keep sending the same response (created), in newer API's we follow deprecation policy and send a 302 (not modified) for now, which the client can pick up and use to display a warning but otherwise not error out.
Then after the deprecation period is up we can return the error.

cpuguy83 commented:
Also, thanks for sticking with this!

yongtang commented:
Thanks @cpuguy83. The PR has been updated with 302 response. Please take a look.

cpuguy83 commented on volume/store/store.go:
```diff
@@ -265,9 +265,11 @@ func (s *VolumeStore) list() ([]volume.Volume, []string, error) {
-func (s *VolumeStore) CreateWithRef(name, driverName, ref string, opts, labels map[string]string) (volume.Volume, error) {
+// In case a volume has already been created with the same name and driver, the already
+// created volumes will be returned
+func (s *VolumeStore) CreateIdempotentWithRef(name, driverName, ref string, opts, labels map[string]string) (volume.Volume, error) {
```
Just a note, we don't need this `Idempotent` change anymore

yongtang replied:
Thanks. `idempotent` has been removed.

cpuguy83 commented:
@yongtang Thanks, I think I like this. We probably need to adjust the responses for backcompat issues, but this is ok for now...Would like to hear what others think.
@tiborvass @aaronlehmann

cpuguy83 commented:
FYI, the reason the behavior is as it is today is for historical reasons (`docker run -v <name>:/foo` is idempotent), but I think ultimately we'd like the volume API to not be.

aaronlehmann commented:
I don't like the difference introduced between `Create` and `CreateWithRef`.

Also, SwarmKit depends on the idempotency of `Create`: https://github.com/docker/docker/blob/master/daemon/cluster/executor/container/adapter.go#L355

Without changes, this use case would be broken.

cpuguy83 commented:
@aaronlehmann indeed that code needs to be moved out of `Create`, these should be the same basic function.

yongtang commented:
Thanks @aaronlehmann @cpuguy83 for the review. The PR has been updated with `Create` in containerAdaptor addressed. The `Create` has also been changed to calling `CreateWithRef` directly. Please take a look.

aaronlehmann commented on daemon/cluster/executor/container/adapter.go:
```diff
@@ -354,10 +355,13 @@ func (c *containerAdapter) createVolumes(ctx context.Context) error {
 		if _, err := c.backend.VolumeCreate(req.Name, req.Driver, req.DriverOpts, req.Labels); err != nil {
-			// TODO(amitshukla): Today, volume create through the engine api does not return an error
-			// when the named volume with the same parameters already exists.
+			// TODO(amitshukla): Today, volume create through the engine api
```
Is there anything remaining to be done? If not, remove the TODO.

aaronlehmann commented:
Unfortunately, needs a rebase.

yongtang commented:
Thanks @aaronlehmann. The PR has been updated and rebased.

AkihiroSuda commented:
Please rebase again?

yongtang commented:
@AkihiroSuda Rebased. Thanks.

LK4D4 commented:
@cpuguy83 @aaronlehmann @AkihiroSuda ok, I'm not sure what's going on :) @yongtang rebased three times, but PR is still in design-review? Did anyone forgot to move it to code-review?

AkihiroSuda commented on integration-cli/docker_cli_volume_test.go:
```diff
@@ -449,3 +452,46 @@ func (s *DockerSuite) TestVolumeCliInspectWithVolumeOpts(c *check.C) {
+	c.Assert(err, check.IsNil)
+	c.Assert(status, check.Equals, http.StatusCreated, check.Commentf(string(b)))
+
+	out, _, err = dockerCmdWithError("volume", "create", "-d", "local", name2, "--opt", "type=tmpfs", "--opt", "o=size=2m,uid=1000")
```
Please add a comment about the intention of L493-L497? (same name with different options?)

Also, I wonder L474-L497 can be split to another test case.

yongtang replied:
Thanks @AkihiroSuda. The PR has been updated with comments added. The test has also been split. Please take a look.

allencloud commented on api/server/router/volume/volume_routes.go:
```diff
@@ -49,6 +51,13 @@ func (v *volumeRouter) postVolumesCreate(ctx context.Context, w http.ResponseWri
+	if volumestore.IsAlreadyExists(err) {
+		version := httputils.VersionFromContext(ctx)
+		if versions.GreaterThan(version, "1.25") {
+			return httputils.WriteJSON(w, http.StatusNotModified, volume)
```
Will here introduce a new API status code of 304?
If that, I am afraid we still need to update the API docs.

yongtang replied:
Thanks @allencloud. The PR has been updated with API docs changed.

allencloud replied:
Oh, thanks a lot. While I am wondering if we need to update the swagger.yml ?

yongtang replied:
@allencloud Updated. Thanks.

allencloud commented on api/swagger.yaml:
```diff
@@ -5904,6 +5904,10 @@ paths:
+        304:
+          description: "The volume has already been created"
+          schema:
+            $ref: "#/definitions/ErrorResponse"
```
Hi, @yongtang Thanks for the update.
I am afraid that I still have some confusion that whether status code 3xx is related to some kinds of error `$ref: "#/definitions/ErrorResponse"`, though I found in other places about 304 in swagger.yml, there is also has an error schema?
I once remember that some time some place you mentioned that in a network framework of programming language, 3xx means no error as well.

yongtang replied:
Ah yes, 3xx only means "additional actions may be required" so typically it is not considered as error for restful APIs. For example, 301 for "Moved Permanently", 307 for "Redirect".

I checked the response it indeed does not return error code. So just updated the API swargger.yml. Thanks

allencloud commented on api/swagger.yaml:
```diff
@@ -5904,6 +5904,8 @@ paths:
           schema:
             $ref: "#/definitions/Volume"
+        304:
+          description: "The volume has already been created"
```
Just mention that, there are another two 304 status codes in swagger.yml where we defined the schema as Error. While I think maybe we should take that into consideration as well.

allencloud commented:
Need a rebase, though. @yongtang 😄

yongtang commented:
@allencloud Rebased. Thanks.

allencloud commented:
Already had much rebase. I am afraid we still need one more ^^ @yongtang

yongtang commented:
Thanks @allencloud. The PR has been rebased.

cpuguy83 commented on api/server/router/volume/volume_routes.go:
```diff
@@ -49,6 +51,13 @@ func (v *volumeRouter) postVolumesCreate(ctx context.Context, w http.ResponseWri
 	volume, err := v.backend.VolumeCreate(req.Name, req.Driver, req.DriverOpts, req.Labels)
+	if volumestore.IsAlreadyExists(err) {
+		version := httputils.VersionFromContext(ctx)
+		if versions.GreaterThan(version, "1.25") {
```
This would need to be "1.26" now.

yongtang replied:
Thanks @cpuguy83. The version has been changed.

cpuguy83 commented on client/volume_create.go:
```diff
@@ -12,6 +14,9 @@ import (
 	var volume types.Volume
 	resp, err := cli.post(ctx, "/volumes/create", nil, options, nil)
+	if resp.statusCode == http.StatusNotModified {
+		return volume, fmt.Errorf("A volume named %s has already been created.", options.Name)
```
I wonder if we should only return an error if options are different (which is something we track now).
This is a breaking CLI change as is.

yongtang committed:
Only return an error if options are different for `docker volume create`

This commit is an update of the last commit such that an error is only returned
if options are different for `docker volume create`.

The change is based on the feedback:
https://github.com/docker/docker/pull/27675#discussion_r100912840

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cpuguy83 for the review. The PR has been updated. Now for `version >= v1.27`, the API will:
1. Return an error if a volume of the same name already exists, and there is a conflict of driver name or driver options.
2. Return `StatusNotModified` if a volume of the same name already exists, and was created with the same driver name and same driver options.

For `version < v1.27`, the old behavior is maintained for backward-compatibility.

Please take a look and let me know if there are any issues.

yongtang commented:
Rebased to fix merge conflict.

thaJeztah commented:
Something seems not right; running `docker volume create foobar` twice;

```bash
root@b68080eb916e:/go/src/github.com/docker/docker# docker volume create foobar
foobar
root@b68080eb916e:/go/src/github.com/docker/docker# docker volume create foobar
(empty line)
```

From the daemon logs;

```
DEBU[0007] Calling GET /_ping
DEBU[0007] Calling POST /v1.29/volumes/create
DEBU[0007] form data: {"Driver":"local","DriverOpts":{},"Labels":{},"Name":"foobar"}
ERRO[0007] Handler for POST /v1.29/volumes/create returned error: http: request method or response status code does not allow body
2017-03-28 15:20:30.450951 I | http: multiple response.WriteHeader calls
```

Also note that 😞 the API version now has to be bumped to v1.29.

vdemeester commented:
/ping @yongtang for taking care of @thaJeztah 👼

yongtang commented:
@vdemeester @thaJeztah Sorry for the late response. I was caught on something else for the past several weeks and probably didn't spend enough time on Docker PRs.

Now I am actively working on outstanding PRs for Docker repo again. There seems to be some changes in error handling in HTTP that makes this PR failing. Hopefully I could address this issue shortly.

yongtang commented:
@vdemeester @thaJeztah The PR has been updated. The issue was that `http.StatusNotModified` does not have a body. This has been fixed now. Also `1.28` has been updated to `1.29`.  Please take a look and let me know if there are any issues.

thaJeztah commented:
The CLI still allows running the command twice, which I think is the intended behavior here;

> So a check of return error type is needed if the caller wants idempotency instead)

```
docker volume create --name dbdata
docker volume create --name dbdata --label foo=bar

docker volume inspect dbdata
[
    {
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/dbdata/_data",
        "Name": "dbdata",
        "Options": {},
        "Scope": "local"
    }
]
```

However, I see errors being logged because returning a body for `created` is not allowed. Not sure what we should do with that;

```
DEBU[0004] Calling GET /_ping
DEBU[0004] Calling GET /_ping
DEBU[0004] Calling POST /v1.29/volumes/create
DEBU[0004] form data: {"Driver":"local","DriverOpts":{},"Labels":{},"Name":"dbdata"}
ERRO[0004] Handler for POST /v1.29/volumes/create returned error: http: request method or response status code does not allow body
2017-04-04 14:41:43.049165 I | http: multiple response.WriteHeader calls
DEBU[0010] Calling GET /_ping
DEBU[0010] Calling GET /_ping
DEBU[0010] Calling POST /v1.29/volumes/create
DEBU[0010] form data: {"Driver":"local","DriverOpts":{},"Labels":{"foo":"bar"},"Name":"dbdata"}
ERRO[0010] Handler for POST /v1.29/volumes/create returned error: http: request method or response status code does not allow body
2017-04-04 14:41:49.288464 I | http: multiple response.WriteHeader calls
```

@cpuguy83 ?

yongtang commented:
Thanks @thaJeztah. I updated the PR. Now `StatusNotModified` will only write the header without the content:
```
                       w.WriteHeader(http.StatusNotModified)
                       return nil
```

The error message should be addressed. Please take a look.

cpuguy83 commented on volume/store/store.go:
```diff
@@ -268,27 +270,45 @@ func (s *VolumeStore) list() ([]volume.Volume, []string, error) {
+// created volumes will be returned together with the errAlreadyExists error.
 func (s *VolumeStore) Create(name, driverName string, opts, labels map[string]string) (volume.Volume, error) {
-	return s.CreateWithRef(name, driverName, "", opts, labels)
+	return s.CreateWithRef(name, driverName, "", opts, labels, func(vol volume.Volume) error {
```
This is the only place we're using this "found" and it seems legit to put this in the volume conflict lookup check.

allencloud commented:
Conflict happened, perhaps a rebase is needed. @yongtang

yongtang committed:
Move found to checkConflict

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
The Jenkins failure are caused by recent changes in other places in Moby. Still looking into the issue.

yongtang commented:
@cpuguy83 The PR has been updated with "found" into volume conflict lookup check. Now all tests finally passes. (`powerpc` test is probably related to Jenkins server. will give it a try again later)

Please take a look and let me know if there are any issues.

vieux commented:
ping @cpuguy83

allencloud commented:
conflict happens, and I am afraid @yongtang needs a bottle of beer and a rebase. 🍺

cpuguy83 commented on volume/store/store.go:
```diff
@@ -268,27 +270,45 @@ func (s *VolumeStore) list() ([]volume.Volume, []string, error) {
 func (s *VolumeStore) Create(name, driverName string, opts, labels map[string]string) (volume.Volume, error) {
-	return s.CreateWithRef(name, driverName, "", opts, labels)
+	return s.CreateWithRef(name, driverName, "", opts, labels, func(vol volume.Volume) error {
+		var options map[string]string
```
Why are we keeping this logic for `CreateWithRef` only?

thaJeztah replied:
ping @yongtang ^^

vdemeester commented:
ping @thaJeztah @yongtang what is the status here ?

AkihiroSuda commented:
ping @yongtang ^^

cpuguy83 commented:
Closing as this has a number of conflicts at this point and hasn't been updated in a long time.
Also see https://github.com/moby/moby/pull/36688 which re-writes much of the volume handling code.
